indexer/categories_holder.cpp

```diff
@@ -195,31 +195,40 @@ void CategoriesHolder::AddCategory(Category & cat, vector<uint32_t> & types)
     shared_ptr<Category> p(new Category());
     p->Swap(cat);
 
-    for (size_t i = 0; i < types.size(); ++i)
-      m_type2cat.insert(make_pair(types[i], p));
+    for (uint32_t const t : types)
+      m_type2cat.insert(make_pair(t, p));
 
-    for (size_t i = 0; i < p->m_synonyms.size(); ++i)
+    for (auto const & synonym : p->m_synonyms)
     {
-      ASSERT(p->m_synonyms[i].m_locale != kUnsupportedLocaleCode, ());
+      auto const locale = synonym.m_locale;
+      ASSERT(locale != kUnsupportedLocaleCode, ());
 
-      StringT const uniName = search::NormalizeAndSimplifyString(p->m_synonyms[i].m_name);
+      auto const uniName = search::NormalizeAndSimplifyString(synonym.m_name);
 
-      vector<StringT> tokens;
+      vector<String> tokens;
       SplitUniString(uniName, MakeBackInsertFunctor(tokens), search::Delimiters());
 
-      for (size_t j = 0; j < tokens.size(); ++j)
-        for (size_t k = 0; k < types.size(); ++k)
-          if (ValidKeyToken(tokens[j]))
-            m_name2type.insert(
-                make_pair(make_pair(p->m_synonyms[i].m_locale, tokens[j]), types[k]));
+      for (auto const & token : tokens)
+      {
+        if (!ValidKeyToken(token))
+          continue;
+        for (uint32_t const t : types)
+        {
+          if (m_name2type.find(locale) == m_name2type.end())
+            m_name2type[locale] = make_unique<Trie>();
```

Comment:
Could you please replace two accesses to map by a single access in the best case and two accesses in the worst case?